x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java

```diff
@@ -50,16 +54,49 @@ protected TypeResolution resolveType() {
         DataType rightType = right().dataType();
         // date math is only possible if one argument is a DATETIME and the other a (foldable) TemporalValue
         if (isDateTimeOrTemporal(leftType) || isDateTimeOrTemporal(rightType)) {
-            if (argumentOfType(DataTypes::isDateTime) == null || argumentOfType(EsqlDataTypes::isTemporalAmount) == null) {
-                return new TypeResolution(
-                    format(null, "[{}] has arguments with incompatible types [{}] and [{}]", symbol(), leftType, rightType)
-                );
+            if (argumentOfType(DataTypes::isDateTime) != null && argumentOfType(EsqlDataTypes::isTemporalAmount) != null) {
+                return TypeResolution.TYPE_RESOLVED;
             }
-            return TypeResolution.TYPE_RESOLVED;
+            if (EsqlDataTypes.isTemporalAmount(leftType) && EsqlDataTypes.isTemporalAmount(rightType)) {
+                return TypeResolution.TYPE_RESOLVED;
+            }
+
+            return new TypeResolution(
+                format(null, "[{}] has arguments with incompatible types [{}] and [{}]", symbol(), leftType, rightType)
+            );
         }
         return super.resolveType();
     }
 
+    @Override
+    public final Object fold() {
+        DataType leftDataType = left().dataType();
+        DataType rightDataType = right().dataType();
+        // TODO: handle time_durations
+        // TODO: handle mixed time_durations and date_periods
+        if (leftDataType == DATE_PERIOD && rightDataType == DATE_PERIOD) {
+            // Both left and right expressions are temporal amounts; we can assume they are both foldable.
```

Comment:
This needs to be extracted and be applicable in block friendly fashion.
Consider `eval m = 1 minute | eval y = 1 year | eval z = m + y` - the code should be in one place regardless if it's a constant (ConstantBlock) or actual backing block.

Reply:
These types can only make sense in an expression involving a DATETIME input and a DATETIME output, since we only have blocks(/vectors) of this type.

Reply:
Yep, what @bpintea said that was my thinking as well. Date math is a bit special as we never want to create blocks of an other type than `DATETIME`. So, the only time when we want to have arithmetics between date periods or time durations is when folding constants.

@costin, I can probably extract this into block friendly code, but I think that will create additional complexity that might not be warranted. Currently, the evaluators for date math are based on the fact that a date time block will be added to a constant value (see[ evaluator definition here](https://github.com/elastic/elasticsearch/blob/aff49dc69a489a40f2aa4a2ae8c31b3c4825b2a8/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Add.java#L87-L89)); so, making the date math logic block based will likely require bigger changes.

If that's okay with you, @costin , I'd prefer not to block-ify the duration/period arithmetics. I think I can put the code for the actual arithmetics into the corresponding operator class, though (`Add.java` and `Sub.java`), then at least all processing code is nicely in one place, even if the code for processing periods/durations does not generate an evaluator.